src/samplers/poisson.jl

```diff
@@ -36,72 +39,85 @@ end
 #
 #   For μ sufficiently large, (i.e. >= 10.0)
 #
-struct PoissonADSampler <: Sampleable{Univariate,Discrete}
-    μ::Float64
-    s::Float64
-    d::Float64
+struct PoissonADSampler{T<:Real} <: Sampleable{Univariate,Discrete}
+    μ::T
+    s::T
+    d::T
     L::Int
 end
 
-PoissonADSampler(μ::Float64) =
-    PoissonADSampler(μ,sqrt(μ),6.0*μ^2,floor(Int,μ-1.1484))
+PoissonADSampler(d::Poisson) = PoissonADSampler(rate(d))
+
+function PoissonADSampler(μ::Real)
+    s = sqrt(μ)
+    d = 6 * μ^2
+    L = floor(Int, μ - 1.1484)
+
+    PoissonADSampler(promote(μ, s, d)..., L)
+end
+
+function rand(rng::AbstractRNG, sampler::PoissonADSampler)
+    μ = sampler.μ
+    s = sampler.s
+    d = sampler.d
+    L = sampler.L
+    μType = typeof(μ)
 
-function rand(rng::AbstractRNG, s::PoissonADSampler)
     # Step N
-    G = s.μ + s.s*randn(rng)
+    G = μ + s * randn(rng, μType)
 
-    if G >= 0.0
-        K = floor(Int,G)
+    if G >= zero(G)
```

Comment:
You can just compare with `0`?

Reply:
I guess that could be used here as well. I think it makes only a different if one works with non-standard number types such as unitful numbers - e.g., `u"1.0m" > 0` errors whereas `u"1.0m" > zero(u"1.0m")` works as expected.

Reply:
Ah, good point, let's be defensive then.